### DISCUSSION TITLE
DirectoryScanJob and IDirectoryScanListener Samples

### DISCUSSION THREAD

**Discussion by gokayokutucu** [quartznet/quartznet#1667] [Q&A]
I haven't found any examples on the internet yet. Are you considering adding any wiki or examples on how DirectoryScanJob works and how IDirectoryScanListener can implement?

**Comment by lahma** [comment]
The documentation is [here](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/quartz-jobs.html#installation) and lacking the examples like you've said. PRs are always welcome.

**Comment by gokayokutucu** [comment]
`IDirectoryScanListener` is implemented like this:

```
public class DeleteOldLogFilesListener : IDirectoryScanListener
    {
        public void FilesUpdatedOrAdded(IReadOnlyCollection<FileInfo> updatedFiles)
        {
            Console.WriteLine("Files were updated");
        }

        public void FilesDeleted(IReadOnlyCollection<FileInfo> deletedFiles)
        {
            Console.WriteLine("Files were deleted");
        }
    }
```
And the configuration is set as follows:

```
        public static IServiceCollection AddQuartzJobs(this IServiceCollection services)
        {
            services.AddSingleton<DirectoryScanJob>();
            services.AddSingleton<DeleteOldLogFilesListener>();


            services.AddQuartz(scheduler =>
            {
                scheduler.ScheduleJob(
                    JobBuilder
                        .Create<DirectoryScanJob>()
                        .WithIdentity("DeleteOldLogFilesJob")
                        .UsingJobData("DIRECTORY_NAME", @"C:\temp\logs")
                        .UsingJobData("DIRECTORY_SCAN_LISTENER_NAME", "DeleteOldLogFilesListener")
                        .Build(),
                    TriggerBuilder
                        .Create()
                        .ForJob("DeleteOldLogFilesJob")
                        .WithSimpleSchedule(s => s
                            .WithIntervalInSeconds(30)
                            .RepeatForever())
                        .Build());
            });

            return services;
        }
```

However, I am not sure to specify a simple scheduler with trigger and the example above doesn't work anyway.

**Comment by gokayokutucu** [comment]
Can someone help me implement this sample please?

**Comment by danielwagn3r** [comment]
> Can someone help me implement this sample please?

Did you get the `DirectoryScanJob` running meanwhile?

**Comment by gokayokutucu** [comment]
> 

Unfortunately, I couldn't run it...

**Comment by desmondrouwhorst** [comment]
Can be me but it took me a while and all the internet, multiple ChatGPT session and my own thoughts, so I will share a version that works for me. hopefully it will help someone.

Programs.cs (web/api/console project):
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddInfrastructureDI(builder.Configuration);


DependencyInjection.cs (infrastructure project) :
private static void AddQuartzDI(
        this IServiceCollection services,
        IConfiguration configuration
    )
    {
        services.Configure<QuartzIOptions>(options => configuration.GetSection("Quartz").Bind(options));

        services.AddQuartzHostedService(options =>
        {
            options.WaitForJobsToComplete = true;
        });

        services.AddSingleton<IScheduler>(provider =>
        {
            var factory = provider.GetRequiredService<ISchedulerFactory>();
            var scheduler = factory.GetScheduler().Result;
            return scheduler;
        });

        services.AddSingleton<DirectoryScanListener>();
        services.AddSingleton<DirectoryScanListenerConfigurator>();

        services.AddQuartz(q =>
        {
            var jobKey = new JobKey("DirectoryScanJob");

            q.AddJob<DirectoryScanJob>(opts => opts.WithIdentity(jobKey));

            q.AddTrigger(opts => opts
                .ForJob(jobKey)
                .WithIdentity("DirectoryScanTrigger")
                .WithSimpleSchedule(x => x.WithIntervalInSeconds(1).RepeatForever())
                .UsingJobData(DirectoryScanJob.DirectoryScanListenerName, "DirectoryScanListener")
                .UsingJobData(DirectoryScanJob.DirectoryProviderName, "DirectoryProvider")
                .StartNow()
            );

            q.AddSchedulerListener<DirectoryScanListenerConfigurator>();
        });
    }
}



DirectoryProvider.cs (infrastructure project):
internal class DirectoryProvider(IOptions<QuartzNetOptions> quartzNetOptions) : IDirectoryProvider
{
    public IReadOnlyList<string> GetDirectoriesToScan(JobDataMap mergedJobDataMap)
    {
        // don't add multiple folder, somehow FilesDeleted get triggers constantly in the DirectoryScanListener
        var dir = new List<string>()
        {
            quartzNetOptions.Value.FolderToWatch
            //@"c:\temp"
        };

        return dir;
    }
}

DirectoryScanListenerConfigurator.cs (infrastructure project):
public class DirectoryScanListenerConfigurator : ISchedulerListener
{
    private readonly IServiceProvider _serviceProvider;

    public DirectoryScanListenerConfigurator(IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
    }

    public Task SchedulerStarted(CancellationToken cancellationToken = default)
    {
        var scheduler = _serviceProvider.GetRequiredService<IScheduler>();
        var listener = _serviceProvider.GetRequiredService<DirectoryScanListener>();

        var options = _serviceProvider.GetRequiredService<IOptions<QuartzNetOptions>>();
        var directoryProvider = new DirectoryProvider(options);

        scheduler.Context.Put("DirectoryProvider", directoryProvider);
        scheduler.Context.Put("DirectoryScanListener", listener);

        return Task.CompletedTask;
    }

make the rest of the functions that are created by the ISchedulerListener return => Task.CompletedTask;



DirectoryScanListener.cs  (infrastructure project):
public class DirectoryScanListener : IDirectoryScanListener 
{
    private readonly ILogger<DirectoryScanListener> _logger;

    public DirectoryScanListener(ILogger<DirectoryScanListener> logger)
    {
        _logger = logger;
    }

    public void FilesDeleted(IReadOnlyCollection<FileInfo> deletedFiles)
    {
        _logger.LogInformation("Found {0} deleted files", deletedFiles.Count);
    }

    public void FilesUpdatedOrAdded(IReadOnlyCollection<FileInfo> updatedFiles)
    {        
        foreach (var file in updatedFiles)
        {
            if (file.Exists)
            {
                _logger.LogInformation("FileName: {fileName}, FilePath: {filePath}, FileExtension {fileExtension}", file.Name, file.FullName, file.Extension);
            }
        }
    }
}


Keep in mind that only newly created files work, only those are pickedup (need to have a creation time newer then the start time of the IDirectoryScanListener)
so copying existing files (created before starting the IDirectoryScanListener) won't get picked up.

When (re)starting the IDirectoryScanListener, then is show the existing file in the folder that is scanned.


If you have questions, just ask.

King regards,
Desmond
